packages/flutter/lib/src/material/snack_bar.dart

```diff
@@ -132,24 +138,106 @@ class _SnackBarActionState extends State<SnackBarAction> {
 
   @override
   Widget build(BuildContext context) {
-    Color? resolveForegroundColor(Set<MaterialState> states) {
-      final SnackBarThemeData snackBarTheme = Theme.of(context).snackBarTheme;
-      if (states.contains(MaterialState.disabled)) {
-        return widget.disabledTextColor ?? snackBarTheme.disabledActionTextColor;
+    final SnackBarThemeData defaults = Theme.of(context).useMaterial3
+        ? _SnackbarDefaultsM3(context)
+        : _SnackbarDefaultsM2(context);
+    final SnackBarThemeData snackBarTheme = Theme.of(context).snackBarTheme;
+
+    MaterialStateColor resolveForegroundColor() {
+      if (widget.textColor is MaterialStateColor) {
+        return widget.textColor! as MaterialStateColor;
       }
-      return widget.textColor ?? snackBarTheme.actionTextColor;
+      return MaterialStateColor.resolveWith((Set<MaterialState> states) {
+        if (states.contains(MaterialState.disabled)) {
+          return widget.disabledTextColor ??
+              snackBarTheme.disabledActionTextColor ??
+              defaults.disabledActionTextColor!;
+        }
+        return widget.textColor ??
+            snackBarTheme.actionTextColor ??
```

Comment:
Can these also be `MaterialStateColor`? If not, it is a little weird that we can use a `MaterialStateColor` for the widget parameter, but not the theme.

Reply:
I'm unsure what you mean - do you mean changing the parameter type? If so, wouldn't that be a breaking change?

Reply:
Sorry, what I was referring to is that we only look at widget.textColor to see if it is a `MaterialStateColor` and resolve it. We should probably do the same for snackBarTheme.actionTextColor and defaults.actionTextColor (i.e. check to see if they are a `MaterialStateColor` and resolve them the same way).

Reply:
Gotcha - thank you for the explanation!